test/e2e_env/universal/meshretry/http.go

```diff
@@ -71,32 +86,117 @@ spec:
 				universal.Cluster, "demo-client", "test-server.mesh",
 			)
 			g.Expect(err).ToNot(HaveOccurred())
-		}).Should(Succeed())
-		Consistently(func(g Gomega) {
-			// --max-time 8 to wait for 8 seconds to beat the default 5s connect timeout
+		}, "10s", "100ms", MustPassRepeatedly(5)).Should(Succeed())
+
+		By("Adding a MeshFaultInjection for test-server")
+		Expect(universal.Cluster.Install(YamlUniversal(meshFaultInjection))).To(Succeed())
+
+		By("Check some errors happen")
+		Eventually(func(g Gomega) {
+			response, err := client.CollectFailure(
+				universal.Cluster, "demo-client", "test-server.mesh",
+				client.NoFail(),
+				client.OutputFormat(`{ "received": { "status": %{response_code} } }`),
+			)
+
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(response.ResponseCode).To(Equal(500))
+		}, "10s", "100ms").Should(Succeed())
+
+		By("Apply a MeshRetry policy")
+		Expect(universal.Cluster.Install(YamlUniversal(meshRetryPolicy))).To(Succeed())
+
+		By("Eventually all requests succeed consistently")
+		Eventually(func(g Gomega) {
```

Comment:
Maybe instead of rely on other policy (`MeshFaultInjection`) you could use new functionality of our test-server (https://github.com/kumahq/kuma/pull/9423, example usage: https://github.com/kumahq/kuma/pull/9437)

Reply:
I don't think it will work better in this case. MeshFaultInjection creates a better setup because, in my experience, failing the 1st request and passing the 2nd one while the header is causing only 'x' consecutive requests to fail. When testing, if we retry and all requests pass, in my opinion, it creates a better setup